transaction/core/src/amount/mod.rs

```diff
@@ -34,31 +54,47 @@ pub struct Amount {
     /// `masked_value = value XOR_8 Blake2B(value_mask | shared_secret)`
     #[prost(fixed64, required, tag = "2")]
     pub masked_value: u64,
+
+    /// `masked_token_id = token_id XOR_8 Blake2B(token_id_mask |
+    /// shared_secret)` 4 bytes long when used, 0 bytes for older amounts
+    /// that don't have this.
+    #[prost(bytes, tag = "3")]
+    #[digestible(omit = "empty")]
+    pub masked_token_id: Vec<u8>,
 }
 
 impl Amount {
     /// Creates a commitment `value*H + blinding*G`, and "masks" the commitment
     /// secrets so that they can be recovered by the recipient.
     ///
     /// # Arguments
-    /// * `value` - The committed value `v`, in picoMOB.
+    /// * `data` - The committed value `v` and token id `i`, in picoMOB.
     /// * `shared_secret` - The shared secret, e.g. `rB` for transaction private
     ///   key `r` and recipient public key `B`.
     #[inline]
-    pub fn new(value: u64, shared_secret: &RistrettoPublic) -> Result<Amount, AmountError> {
+    pub fn new(data: AmountData, shared_secret: &RistrettoPublic) -> Result<Amount, AmountError> {
         // The blinding is `Blake2B("blinding" | shared_secret)`
         let blinding: Scalar = get_blinding(shared_secret);
 
-        // Pedersen commitment `v*H + b*G`.
-        let commitment = CompressedCommitment::new(value, blinding);
+        // Pedersen generators
+        let generator = generators(*data.token_id);
+
+        // Pedersen commitment `v*H_i + b*G`.
+        let commitment = CompressedCommitment::new(data.value, blinding, &generator);
+
+        // The value is XORed with the first 8 bytes of the mask.
+        // `v XOR_8 Blake2B(value_mask | shared_secret)`
+        let masked_value: u64 = data.value ^ get_value_mask(shared_secret);
 
         // The value is XORed with the first 8 bytes of the mask.
```

Comment:
This comment looks copy-pasted. Should be something like "The token_id is XORed...", no?

Reply:
yes thank you